### ISSUE TITLE
Occasional JS redraw error in CI

### ISSUE THREAD

**Issue by noahgibbs** [scarpe-team/scarpe#468]
I've seen it twice now, but this time I got failure logs.

The error happens in test_drawables.rb in test_hide_show. These logs have it happening with the Progress bar, though I don't think it's specific to the progress bar. There are a few odd things in the log.

The immediate reported problem is this:

~~~
[894] ERROR Webview::WebWrangler::DOMWrangler: Could not complete JS redraw! /Users/runner/work/scarpe/scarpe/lib/scarpe/wv/web_wrangler.rb:692:in `full_message': JS runtime error: "null is not an object (evaluating 'document.getElementById(\"14\").style')"! (Scarpe::JSRuntimeError)
~~~

So it's trying to grab object ID 14 (the progress bar) and get its style, but it's not there. This is happening after shutdown has already started. For instance, you can see a heartbeat being cancelled a bit above the error due to shutdown. It looks like a *lot* of setting display:none on a lot of different widgets happens after shutdown. So: not sure why shutdown happened so soon. But shutdown should certainly be cancelling redraws, because they're likely to fail after shutdown happens. But that also means shutdown was happening before we did all the assertions. Not great.

Weirdly, it *also* looks like the progress bar didn't have its HTML ID in its div. Here's a chunk of the JS that created the DOM tree:

```
<div id=\\\"9\\\" style=\\\"display:flex;flex-direction:row;flex-wrap:wrap;align-content:flex-start;justify-content:flex-start;align-items:flex-start;width:100%\\\"></div><img id=\\\"10\\\" src=\\\"https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png\\\" /><div id=\\\"11\\\" style=\\\"left:0px;top:0px\\\"><svg width=\\\"100\\\" height=\\\"100\\\"><line x1=\\\"0\\\" y1=\\\"0\\\" x2=\\\"100\\\" y2=\\\"100\\\" style=\\\"stroke:black;stroke-width:4\\\"></line></svg></div><select id=\\\"12\\\" onchange=\\\"scarpeHandler('12-change', this.options[this.selectedIndex].value)\\\"><option value=\\\"A\\\">A</option><option value=\\\"B\\\">B</option></select><p id=\\\"13\\\" style=\\\"font-size:12px\\\">Hello</p><div class=\\\"progress\\\" style=\\\"width: 90%\\\"><div class=\\\"progress-bar progress-bar-striped progress-bar-animated\\\" role=\\\"progressbar\\\" aria-valuenow=\\\"0.0\\\" aria-valuemin=\\\"0\\\" aria-valuemax=\\\"100\\\" style=\\\"width: 0.0%\\\"></div></div><input type=\\\"radio\\\" id=\\\"15\\\" onclick=\\\"scarpeHandler('15-click')\\\" name=\\\"ooga\\\" />
```

If you look closely (may be easier in the attached logs), you can find ID 13 and 15, but no ID 14. So the query may be failing because of that, too. But the progress bar should be rendering with the ID just fine, looking at the Calzini code. This test normally passes. There *is* a progress bar there, just with no HTML ID.

A lot of that test uses ".to_html" for checking. None of the .to_html calls will directly run Javascript. They just ask the Ruby drawable to render themselves to local HTML in a string. So maybe the answer to "why did it shut down so soon?" is that we did all the Ruby stuff in the assertions, they succeeded, and we didn't need to get anything else from JS. The Ruby Fiber will cheerfully finish and exit if it's convinced it's finished successfully, even if there's JS still hanging around (e.g. a redraw.) Usually that's a good thing. The last JS that the test waits for is: assert_includes dom_html, "display: none". I wouldn't expect that to log, though.

I'm pretty sure that the failing part, querying HTML ID 14's style, is when we're trying to hide it. If it was rendered without an HTML ID, that would do it. But why was it rendered without an HTML ID? That's weird.

Basically, it's weird that the terminate is happening so early. Ruby won't wait after telling everything to hide, but it *should* wait for the "dom_html" call in that assertion, which shouldn't run until all the hiding has finished.

If I just turn off redraw on terminate, we won't get that redraw error. I hope we'll still get an error, though (or a hang?), because I *think* the terminate is happening before all the assertions have finished.

[test failure logs - js redraw failed.zip](https://github.com/scarpe-team/scarpe/files/13487464/test.failure.logs.-.js.redraw.failed.zip)

**Comment by noahgibbs** [comment]
Possibly related: I've seen at least once another (inconsistent) error in head-of-main on the same test:

```
Failure:
TestDrawables#test_hide_show [(eval):10]
Minitest::Assertion: expected drawable Scarpe::ShoesSpecProxy to be hidden!
```

**Comment by noahgibbs** [comment]
The progress ID thing was a Tiranti error, fixed now.
